03-attribute-operations.Rmd

```diff
@@ -499,14 +507,211 @@ world_data = world %>% st_set_geometry(NULL)
+recl = reclassify(r, rcl = rcl)
+```
+
+Raster algebra is another classical use case of local operations.
+This includes adding, subtracting and squaring two rasters.
+Raster algebra also allows logical operations such as finding all raster cells that are greater than a specific value (5 in our example below).
+The **raster** package allows all these operations in a way natural to R users.
+Please see the `Raster`-vignette for a more detailed description on algebraic operations (`vignette("Raster", package = "raster")`).
+
+```{r, eval = FALSE}
+r + r
+r^2
+log(r)
+r > 5
+```
+
+The calculation of the normalized difference vegetation index (NDVI) is one of the most famous local, i.e. pixel-by-pixel, raster operations.
+It ranges between - 1 and 1 with positive values indicating the presence of living plants (mostly > 0.2).
+To calculate the NDVI, one uses the red and near-infrared bands of remotely sensed imagery (e.g., Landsat or Sentinel imagery) exploiting the fact that vegetation absorpts light heavily in the visible light spectrum, and especially in the red channel, while reflecting it in the near-infrared spectrum.
+
+$$
+\begin{split}
+NDVI&= \frac{\text{NIR} - \text{Red}}{\text{NIR} + \text{Red}}\\
+\end{split}
+$$
+where NIR = near infrared channel
+      Red = red channel
+
+Predictive mapping is another interesting application of local raster operations.
+The response variable correspond to measured or observed points in space, for example, species richness, the presence of landslides, tree disease or crop yield.
+Consequently, we can easily retrieve space- or airborne predictor variables from various rasters (elevation, pH, precipitation, temperature, landcover, soil class, etc.).
+Subsequently, we model our response as a function of our predictors using `lm`, `glm`, `gam` or a machine-learning technique. 
+To make a spatial prediction, all we have to do, is to apply the estimated coefficients to the predictor rasters, and summing up the resulting output rasters (Chapter ??; see also @muenchow_predictive_2013).
+<!-- add reference to chapter ecological modeling -->
+
+While local functions operate on one cell, though possibly from multiple layers, **focal** operations take into account a central cell and its neigbors.
+The neigborhood (also named kernel, filter or moving window) under consideration is typically of size 3-by-3 cells (that is the central cell and its eight surrounding neighbors) but can take on any other (not necessarily rectangular) shape as defined by the user.
+A focal operation applies an aggregation function to all cells within the specified neighborhood, uses the corresponding output as the new value for the the central cell, and moves on to the next central cell (Figure @ref(fig:focal_example)).
+Other names for this operation are spatial filtering and convolution [@burrough_principles_2015].
+
+In R, we can use the `focal()` function to perform spatial filtering. 
```

Comment:
Good description of focal operations (are they spatial or attribute thought ; ) I think they are fine here but could also fit in Chapter 4. Let me know when your teeth into that chapter if you think it fits...

Reply:
Yes, we could add them to Chapter 4 as well. In any case, I would recommend not to spread the different pillars of map algebra into different chapters, i.e. if we put the focal stuff into c4 we should do so with the local, zonal and global operations. I guess, this is also what most readers would expect. 
Thanks for adding the raster question to the list for the reviewers. Good idea to move on in the meantime 😃